### PR TITLE
fix: rewrite ifnull and nvl as coalesce

### LINKED ISSUE #14858: bug: The results of the NVL function may be incorrect
### Search before asking

- [X] I had searched in the [issues](https://github.com/datafuselabs/databend/issues) and found no similar issues.

### Version

main branch

### What's Wrong?

because `nvl` rewrite to `if(is_not_null(x), x, y)`, so `x` Expr will calculate twice,  if `x` is volatile like `nullif(rand()>0.5, true)`, The results of the two calculations will be different，example, below sql should be always false, but this bug will make it possible to be true

`select nvl(nullif(rand()>0.5, true), false) is null from (unnest(range(1,10)));`
```
root@localhost:8000/default> select nvl(nullif(rand()>0.5, true), false) is null from (unnest(range(1,10)));

SELECT
  nvl(nullif(rand() > 0.5, TRUE), false) IS NULL
FROM
  (unnest(RANGE(1, 10)))

┌──────────────────────────────────────────────────┐
│ nvl(nullif((rand() > 0.5), true), false) is null │
│                      Boolean                     │
├──────────────────────────────────────────────────┤
│ true                                             │
│ false                                            │
│ false                                            │
│ false                                            │
│ false                                            │
│ false                                            │
│ true                                             │
│ false                                            │
│ true                                             │
└──────────────────────────────────────────────────┘
9 rows read in 0.110 sec. Processed 1 rows, 1 B (9.06 rows/s, 9 B/s)
```

### How to Reproduce?

run sql: select nvl(nullif(rand()>0.5, true), false) is null from (unnest(range(1,10)));

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by PsiACE**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

I think we can simply rewrite ifnull and nvl as coalesce to fix this problem.

I noticed that DuckDB does it this way.

image

- Fixes #14858

## Tests

- [ ] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test  - _Explain why_

## Type of change

- [x] Bug Fix (non-breaking change which fixes an issue)
- [ ] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/14877)

### CODE REVIEW

**Review by sundy-li — APPROVED** [review]

File: src/query/sql/src/planner/semantic/type_check.rs [hunk 1]
```diff
@@ -2620,6 +2604,14 @@ impl<'a> TypeChecker<'a> {
...
                     lit: Literal::Null,
                 });
+                new_args.push(Expr::Literal {
+                    span,
+                    lit: Literal::Null,
+                });
+                new_args.push(Expr::Literal {
+                    span,
+                    lit: Literal::Null,
+                });
```
Comment (PsiACE) [inline src/query/sql/src/planner/semantic/type_check.rs:2614]: This is a workaround that ensures there are always at least 3 args for the if function to work properly.

Comment (PsiACE) [reply src/query/sql/src/planner/semantic/type_check.rs:2614]: before:

```
select coalesce(null, null);
error: APIError: ResponseError with 1065: error:
  --> SQL:1:8
  |
1 | select coalesce(null, null)
  |        ^^^^^^^^^^^^^^^^^^^^ no overload satisfies `if(NULL)`
```

after:

```
select coalesce(null, null);

SELECT
  coalesce(NULL, NULL)

┌──────────────────────┐
│ coalesce(null, null) │
│         Null         │
├──────────────────────┤
│ NULL                 │
└──────────────────────┘
```

Comment (PsiACE) [reply src/query/sql/src/planner/semantic/type_check.rs:2614]: image

**Comment by andylokandy** [comment]
can `coalesce` resist  `rand() > 0.5`?

**Comment by PsiACE** [comment]
> can `coalesce` resist `rand() > 0.5`?

From the results of the test, yes. Our implementation of `coalesce` is correct.

**Comment by PsiACE** [comment]
> can `coalesce` resist `rand() > 0.5`?

I tested another example, unfortunately, it cannot. I have examined the implementation of DuckDB and it also cannot.

**Comment by PsiACE** [comment]
This PR should be a relatively reasonable fix. Would you like to review it again? @sundy-li

**Review by sundy-li — APPROVED** [review]

**Pull request merged**